src/transform/src/predicate_pushdown.rs

```diff
@@ -84,357 +82,408 @@ impl PredicatePushdown {
         relation: &mut RelationExpr,
         get_predicates: &mut HashMap<Id, HashSet<ScalarExpr>>,
     ) {
-        relation.visit_mut_pre(&mut |e| {
-            // TODO(#2592): we want to replace everything inside the braces
-            // with the single line below
-            // `self.action(e, &mut get_predicates);`
-            // This is so that you have a series of dependent views
-            // A->B->C, you want to push propagated filters
-            // from A all the way past B to C if possible.
-            // Before this replacement can be done, we need to figure out
-            // replanning joins after the new predicates are pushed down.
-            if let RelationExpr::Filter { input, predicates } = e {
+        // TODO(#2592): we want to replace everything inside the braces
+        // with the single line below
+        // `self.action(e, &mut get_predicates);`
+        // This is so that you have a series of dependent views
+        // A->B->C, you want to push propagated filters
+        // from A all the way past B to C if possible.
+        // Before this replacement can be done, we need to figure out
+        // replanning joins after the new predicates are pushed down.
+        match relation {
+            RelationExpr::Filter { input, predicates } => {
                 if let RelationExpr::Get { id, .. } = **input {
                     // We can report the predicates upward in `get_predicates`,
                     // but we are not yet able to delete them from the `Filter`.
                     get_predicates
                         .entry(id)
                         .or_insert_with(|| predicates.iter().cloned().collect())
                         .retain(|p| predicates.contains(p));
+                } else {
+                    input.visit1_mut(|e| self.dataflow_transform(e, get_predicates));
```

Comment:
Follow up from my earlier question: and we'll only be trying to apply `PredicatePushdown` to the children of `RelationExprs` who have `PredicatePushdown` applied to them?

Reply:
I think that statement is correct, but I might be missing some angle. If all you are saying is "we are using explicit recursive descent now, rather than letting `visit_mut` handle things" then 100% agree.